plugins/cache/bap_cache.ml

```diff
@@ -0,0 +1,94 @@
+open Core_kernel
+open Regular.Std
+open Bap.Std
+open Bap_cache_types
+
+include Self ()
+
+module Filename = Caml.Filename
+module Utils = Bap_cache_utils
+module Cfg = Bap_cache_config
+
+let (//) = Filename.concat
+
+let cache_dir = Cfg.cache_dir
+let cache_data = Cfg.cache_data
+
+module Upgrade = struct
+
+  let index_versions = [2;1]
+  let index_file v = sprintf "index.%d" v
+  let index_files = List.map index_versions ~f:index_file
+
+  let find_index () =
+    let files = List.map index_files ~f:(fun x -> cache_dir () // x) in
+    List.find files ~f:Sys.file_exists
+
+  let get_version path =
+    let file = Filename.basename path in
+    match String.chop_prefix file "index." with
+    | None -> Ok 1
+    | Some v ->
+      try Ok (int_of_string v)
+      with _ ->
+        Error (Error.of_string (sprintf "unknown version %s" v))
+
+  let rename from to_ =
+    Sys.rename from to_;
+    Unix.chmod to_ 0o444
+
+  let upgrade_from_index_v2 file =
+    let open Compatibility.V2 in
+    try
+      let idx = Utils.from_file (module Compatibility.V2) file in
+      let dir = cache_data () in
+      Map.iteri idx.entries ~f:(fun ~key ~data:{path} ->
+          rename path (dir // Data.Cache.Digest.to_string key))
+    with e ->
+      warning "can't read entries from index version 2: %s"
+        (Exn.to_string e)
+
+  let from_index () = match find_index () with
+    | None -> ()
+    | Some file ->
+      Cfg.(write default);
+      match get_version file with
+      | Ok 2 ->
+        upgrade_from_index_v2 file;
+        Sys.remove file
+      | Ok ver ->
+        warning "can't read entries from index version %d" ver;
```

Comment:
An unsupported version is basically the same as an unknown version, we can't upgrade from any, so either both should be a warning or both should be an error. I suggest turning any unsupported/unknown case into a warning, but cleanup the old index, so that this warning is not repeated every time.